pkg/controller/serviceexternalip/controller_test.go

```diff
@@ -0,0 +1,307 @@
+	_, err = controller.client.CoreV1().Services(service.Namespace).Update(ctx, service, metav1.UpdateOptions{})
+	assert.NoError(t, err)
+	checkForServiceExternalIP(t, controller, ctx, service.Name, service.Namespace, "1.2.4.4")
+	checkExternalIPPoolUsed(t, controller, ctx, "eip1", 0)
+	checkExternalIPPoolUsed(t, controller, ctx, "eip2", 1)
+
+	// Specify non-existent IP of ExternalIPPool.
+	service.Spec.LoadBalancerIP = "1.2.4.6"
+	service.Annotations[antreaagenttypes.ServiceExternalIPPoolAnnotationKey] = "eip2"
+	_, err = controller.client.CoreV1().Services(service.Namespace).Update(ctx, service, metav1.UpdateOptions{})
+	assert.NoError(t, err)
+	checkForServiceExternalIP(t, controller, ctx, service.Name, service.Namespace, "")
+	checkExternalIPPoolUsed(t, controller, ctx, "eip1", 0)
+	checkExternalIPPoolUsed(t, controller, ctx, "eip2", 0)
+
+	// Change Service type.
+	service.Spec.LoadBalancerIP = ""
+	_, err = controller.client.CoreV1().Services(service.Namespace).Update(ctx, service, metav1.UpdateOptions{})
+	assert.NoError(t, err)
+	checkForServiceExternalIP(t, controller, ctx, service.Name, service.Namespace, "1.2.4.4")
+
+	service.Spec.Type = corev1.ServiceTypeClusterIP
+	_, err = controller.client.CoreV1().Services(service.Namespace).Update(ctx, service, metav1.UpdateOptions{})
+	assert.NoError(t, err)
+	checkForServiceExternalIP(t, controller, ctx, service.Name, service.Namespace, "")
+	checkExternalIPPoolUsed(t, controller, ctx, "eip1", 0)
+	checkExternalIPPoolUsed(t, controller, ctx, "eip2", 0)
+
+	// Change Service type back.
+	service.Spec.Type = corev1.ServiceTypeLoadBalancer
+	_, err = controller.client.CoreV1().Services(service.Namespace).Update(ctx, service, metav1.UpdateOptions{})
+	assert.NoError(t, err)
+	checkForServiceExternalIP(t, controller, ctx, service.Name, service.Namespace, "1.2.4.4")
+	checkExternalIPPoolUsed(t, controller, ctx, "eip2", 1)
+
+	// Delete Service.
+	err = controller.client.CoreV1().Services(service.Namespace).Delete(ctx, service.Name, metav1.DeleteOptions{})
+	assert.NoError(t, err)
+	checkExternalIPPoolUsed(t, controller, ctx, "eip2", 0)
+}
+
+func checkForServiceExternalIP(t *testing.T, controller *loadBalancerController, ctx context.Context, name, namespace, expectedExternalIP string) {
+	assert.Eventually(t, func() bool {
+		serviceUpdated, err := controller.client.CoreV1().Services(namespace).Get(ctx, name, metav1.GetOptions{})
+		require.NoError(t, err)
+		externalIP := getServiceExternalIP(serviceUpdated)
+		return externalIP == expectedExternalIP
+	}, 500*time.Millisecond, 100*time.Millisecond)
+}
+
+func checkExternalIPPoolUsed(t *testing.T, controller *loadBalancerController, ctx context.Context, poolName string, used int) {
+	exists := controller.externalIPAllocator.IPPoolExists(poolName)
+	require.True(t, exists)
+	assert.Eventually(t, func() bool {
+		eip, err := controller.crdClient.CrdV1alpha2().ExternalIPPools().Get(context.TODO(), poolName, metav1.GetOptions{})
```

Comment:
It passes a ctx but creates a new context.
But I feel no need to create the cancelable context in this test file, as there is no external call and all test code are executed sequentially.

Reply:
deleted the cancelable context in this test file.